File: Stripe/STPSourceWeChatPayDetails.m
//
//  STPSourceWeChatPayDetails.m
//  StripeiOS
//
//  Created by Yuki Tokuhiro on 6/4/19.
//  Copyright © 2019 Stripe, Inc. All rights reserved.
//

#import "STPSourceWeChatPayDetails.h"

#import "NSDictionary+Stripe.h"

@interface STPSourceWeChatPayDetails()
@property (nonatomic, copy, readwrite) NSString *weChatAppURL;
@property (nonatomic, copy, readwrite) NSDictionary *allResponseFields;
@end

@implementation STPSourceWeChatPayDetails


#pragma mark - Description

- (NSString *)description {
    NSArray *props = @[
                       [NSString stringWithFormat:@"%@: %p", NSStringFromClass([self class]), self],
                       [NSString stringWithFormat:@"weChatAppURL = %@", self.weChatAppURL],
                       ];
    
    return [NSString stringWithFormat:@"<%@>", [props componentsJoinedByString:@"; "]];
}

#pragma mark - STPAPIResponseDecodable

+ (nullable instancetype)decodedObjectFromAPIResponse:(nullable NSDictionary *)response {
    NSDictionary *dict = [response stp_dictionaryByRemovingNulls];
    if (!dict) {
        return nil;
    }
    STPSourceWeChatPayDetails *details = [[self class] new];
    details.weChatAppURL = [dict stp_stringForKey:@"ios_native_url"];
    details.allResponseFields = dict;
    return details;
}

@end

File: StripePayments/StripePayments/API Bindings/Models/Sources/Types/STPSourceWeChatPayDetails.swift
//
//  STPSourceWeChatPayDetails.swift
//  StripePayments
//
//  Created by Yuki Tokuhiro on 6/4/19.
//  Copyright © 2019 Stripe, Inc. All rights reserved.
//

import Foundation

/// Details of a WeChat Pay Source.
public class STPSourceWeChatPayDetails: NSObject, STPAPIResponseDecodable {
    /// A URL to the WeChat App.
    /// Use `STPRedirectContext` instead of redirecting users yourself.
    @objc public private(set) var weChatAppURL: String?
    @objc public private(set) var allResponseFields: [AnyHashable: Any] = [:]

    // MARK: - Description
    /// :nodoc:
    @objc public override var description: String {
        let props = [
            String(format: "%@: %p", NSStringFromClass(STPSourceWeChatPayDetails.self), self),
            "weChatAppURL = \(weChatAppURL ?? "")",
        ]

        return "<\(props.joined(separator: "; "))>"
    }

    // MARK: - STPAPIResponseDecodable
    public class func decodedObject(fromAPIResponse response: [AnyHashable: Any]?) -> Self? {
        guard let response = response else {
            return nil
        }
        let dict = response.stp_dictionaryByRemovingNulls()
        let details = self.init()
        details.weChatAppURL = dict.stp_string(forKey: "ios_native_url")
        details.allResponseFields = response
        return details
    }

    override required init() {
        super.init()
    }
}
